### PR TITLE
fixes #28 and #32

### LINKED ISSUE #28: boolean Field gets returned as null instead of false
I use typegql with objection.js model, on this model I have a Field like this:

```
@ObjectType
class WorkspaceModel extends TimestampedModel {
  static tableName = 'workspaces'
  @Field() readonly id!: number
  @Field() user_id: number
  @Field() name: string
  @Field()
  is_open: boolean
}
```

This is the WorkspaceModel object I am returning in my `Query` :
```
{
      "id": 3,
      "is_open": false,
      "name": "Workspace from test",
      "openEnrolmentCount": 0,
      "role": "OWNER",
      "topicsCount": 0,
    }
```

this is what I get from graphql:
```
{
      "id": 3,
      "is_open": null,
      "name": "Workspace from test",
      "openEnrolmentCount": 0,
      "role": "OWNER",
      "topicsCount": 0,
    }
```
So I ran a test in my boilerplate an indeed using this schema:

```javascript
@ObjectType({ description: 'Simple product object type' })
class Product extends Model {
  @Field() is_open: boolean
}

@SchemaRoot()
class SuperSchema {
  @Query({ type: Product, isNullable: false })
  prod(): Product {
    const product = new Product()
    product.is_open = false
    return product
  }
}

```

image

when the field is `true` it is returned correctly.

### PR DESCRIPTION
**Pull request by capaj**

### CODE REVIEW

File: src/domains/field/compiler/resolver.ts [hunk 1]
```diff
@@ -62,6 +62,7 @@ function computeFinalArgs(
 export function compileFieldResolver(
   target: Function,
   fieldName: string,
+  isGetter: boolean,
```
Comment (pie6k) [inline src/domains/field/compiler/resolver.ts]: Can it be replaced in favor of using `Object.getOwnPropertyDescriptor` to detect it automatically? Passing it over like this might make it harder to maintain

File: src/domains/field/compiler/resolver.ts [hunk 2]
```diff
@@ -70,8 +71,18 @@ export function compileFieldResolver(
 
   return async (source: any, args = null, context = null, info = null) => {
     await performHooksExecution(beforeHooks, source, args, context, info);
-    const instanceField =
-      (source && source[fieldName]) || target.prototype[fieldName];
+
+    if (isGetter) {
```
Comment (pie6k) [inline src/domains/field/compiler/resolver.ts]: This whole block is hard to understand for me. Could we make it external function that would abstract it away with proper name. It seems like the point here is to get target property.

File: src/domains/field/index.ts [hunk 3]
```diff
@@ -25,7 +25,13 @@ export function Field(options?: FieldOptions): PropertyDecorator {
       ...options,
     };
 
+    const descriptor = Object.getOwnPropertyDescriptor(
+      targetInstance,
+      fieldName,
+    );
+
     fieldsRegistry.set(targetInstance.constructor, fieldName, {
```
Comment (pie6k) [inline src/domains/field/index.ts]: If it can be calculated from `targetInstance`, then passing it as another parameter is redundant in my opinion.

File: src/domains/field/registry.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ export interface FieldInnerConfig {
   property: string;
   description?: string;
   isNullable?: boolean;
+  isGetter?: boolean;
```
Comment (pie6k) [inline src/domains/field/registry.ts]: I'm quite strongly against having it explicitly here. It looks like you might define it. But it's part of business logic that could propably be calculated 'on the fly' when needed.

Config is what user or system want it to be. Getter is implementation detail, so I'd say it's not part of `config`.

File: src/services/utils/gql/types/parseNative.ts [hunk 5]
```diff
@@ -31,10 +31,15 @@ export function inferTypeByTarget(target: Function, key?: string) {
   }
 
   const returnType = Reflect.getMetadata('design:returntype', target, key);
+
   if (returnType) {
     return returnType;
   }
 
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
```
Comment (pie6k) [inline src/services/utils/gql/types/parseNative.ts]: I think this function is already quite complicated and not obvious. Making external function like `isGetter` would make it more readable and obvious about what is the point here.

File: src/test/field/getters.spec.ts [hunk 6]
```diff
@@ -3,16 +3,25 @@ import { ObjectType, Field, compileObjectType } from '~/domains';
 
 describe('Fields based on getters', () => {
   it('Will work with getter field', async () => {
+    let foo: Foo;
     @ObjectType({ description: 'Simple product object type' })
     class Foo {
+      a: number;
       @Field()
       get bar(): number {
-        return 42;
```
Comment (pie6k) [inline src/test/field/getters.spec.ts]: I don't think it's good way to perform tests by throwing errors to stop them.

There might be different ways to accomplish this like adding instance field and then returning result like `return this.foo` and that should be enough to test if instance was avaliable

File: src/test/field/index.spec.ts [hunk 7]
```diff
@@ -14,11 +14,31 @@ describe('Field', () => {
     @ObjectType()
     class Foo {
       @Field() bar: string = 'baz';
+      @Field() undef: boolean = undefined;
```
Comment (pie6k) [inline src/test/field/index.spec.ts]: I'd say this test is out of scope of `resolving default value`. It's rather another test case like `Resolves with edge-case default values`

**Comment by pie6k** [comment]
Thanks for this PR! I've added my code review :)

File: src/domains/field/compiler/resolver.ts (hunk 1 above)
Comment (capaj) [reply src/domains/field/compiler/resolver.ts]: it could, but doing `Object.getOwnPropertyDescriptor` is performance costly. This might sound like a micro optimisation, but it's not. See for example https://github.com/chaijs/chai/issues/127#issuecomment-12210969.

Having it detected only once is worth the faster performance IMHO. Decorators always run statically at startup, so it doesn't make sense to support a scenario where someone would change a property into a getter at runtime.

File: src/domains/field/registry.ts (hunk 4 above)
Comment (capaj) [reply src/domains/field/registry.ts]: like I've mentioned-detecting getter always using `Object.getOwnPropertyDescriptor` is quite costly.

**Comment by pie6k** [comment]
@capaj I've played a bit with it and actually it could be solved in much simpler way: I've created https://github.com/prismake/typegql/pull/37/files - it seems to solve both #28 and #32 (check added test cases) too.

**Comment by capaj** [comment]
@pie6k this only fixes the #32, not the #28 or am I missing something?

**Comment by capaj** [comment]
@pie6k also it doesn't fix the context-in the test you're defining the property on the prototype. We want the instance as the context, not the prototype.

**Comment by pie6k** [comment]
Could you show me code of some failing test for that?

**Comment by pie6k** [comment]
About #28 - https://github.com/prismake/typegql/pull/38 fixes that. Again it's way simpler than here because it doesnt include getters workaround there

**Comment by pie6k** [comment]
@capaj - I've updated context test to use both prototype and instance context. It's still passing. Let me know if I've missed something you mean :)

**Comment by capaj** [comment]
@pie6k yeah seems to be working as expected. Good that it doesn't have to do those costly `Object.getOwnPropertyDescriptor` calls. :+1:
